### PR TITLE
Update optimizer for num_steps_per_communication

### PR DESCRIPTION
**Pull request by hanbinhu**

### CODE REVIEW

File: bluefog/torch/optimizers.py [hunk 1]
```diff
@@ -312,8 +318,16 @@ def hook(module, *unused):
                     # End forward computation timeline
                     bf.timeline_end_activity(parent_name+'.'+name)
                 if p.requires_grad:
-                    handle = self._neighbor_allreduce_data_async(p)
-                    self._handles[p] = handle
+                    if self._neighbor_allreduce_delay[p] <= 0:
+                        raise AssertionError(
+                            "Unexpected behavior: forward computation were computed "
+                            "more than num_steps_per_communication times before call "
+                            "to step(). Adjust num_steps_per_communication to "
```
Comment (ybc1991) [inline bluefog/torch/optimizers.py:325]: 我们现在这样的写法是不是不会出现这样的情况？

File: bluefog/torch/optimizers.py [hunk 2]
```diff
@@ -436,10 +444,18 @@ def hook(module, *unused):
...
-                    handle = bf.win_put_nonblocking(
-                        tensor=p.data, name=parent_name+'.'+name,
-                        dst_weights=self.dst_weights, require_mutex=False)
-                    self._handles[p] = handle
+                    if self._bluefog_delay[p] <= 0:
+                        raise AssertionError(
+                            "Gradients were computed more than "
+                            "backward_passes_per_step times before call "
+                            "to step(). Increase backward_passes_per_step to "
+                            "accumulate gradients locally.")
```
Comment (ybc1991) [inline bluefog/torch/optimizers.py:452]: same comment

File: bluefog/torch/optimizers.py [hunk 3]
```diff
@@ -169,8 +172,16 @@ def _register_hooks(self):
     def _make_hook(self, p):
         def hook(*ignore):
             assert not p.grad.requires_grad
-            handle = self._allreduce_grad_async(p)
-            self._handles[p] = handle
+            if self._allreduce_delay[p] <= 0:
+                raise AssertionError(
+                    "Unexpected behavior: forward computation were computed "
```
Comment (ybc1991) [inline bluefog/torch/optimizers.py:177]: 我们allreduce 还是backward的

File: bluefog/torch/optimizers.py [hunk 4]
```diff
@@ -312,8 +318,16 @@ def hook(module, *unused):
...
-                    handle = self._neighbor_allreduce_data_async(p)
-                    self._handles[p] = handle
+                    if self._neighbor_allreduce_delay[p] <= 0:
+                        raise AssertionError(
+                            "Unexpected behavior: forward computation were computed "
+                            "more than num_steps_per_communication times before call "
+                            "to step(). Adjust num_steps_per_communication to "
+                            "accumulate gradients locally.")
+                    self._neighbor_allreduce_delay[p] -= 1
+                    if self._neighbor_allreduce_delay[p] == 0:
```
Comment (ybc1991) [inline bluefog/torch/optimizers.py:328]: 我在想和 horovod backwards做法不一样， 我们应该是第一次forward的时候触发 neighbor allreduce 然后一直不触发？

File: bluefog/torch/optimizers.py [hunk 5]
```diff
@@ -337,19 +351,10 @@ def turn_off_timeline(self):
...
-
-        for p, handle in self._handles.items():
-            if handle is None:
-                handle = self._neighbor_allreduce_data_async(p)
-                self._handles[p] = handle
-
         with torch.no_grad():
             for p, handle in self._handles.items():
                 output = bf.synchronize(handle)
+                self._neighbor_allreduce_delay[p] = self._num_steps_per_communication
```
Comment (ybc1991) [inline bluefog/torch/optimizers.py:357]: 如果我没有理解错的话， 两种场景做法
1.  w_{i+1} = neighbor_allredce(w_i) + \sum \grad_j(w_i)  ---- multiple gradients with different data but same iter w_i,

2. w_{i, j+1} = w_{i, j} + \sum \grad_j(w_{i,j}),  j = 0, 1, ....J-1
    w_{i+1, 0} = neighbor_allredce(w_{i, 0}）
你这里实现的还是1？

**Review by ybc1991 — CHANGES_REQUESTED** [review]

File: bluefog/torch/optimizers.py (hunk 5 above)
Comment (ybc1991) [reply bluefog/torch/optimizers.py:357]: 我觉得我理解错了， 你实现的确实是2.

如果要实现1， 是不是直接就是

for j in range(J):
      output = model(data_batch_i)
      loss = xxx
      loss.backward()
opt.step()

而 2 就是
opt.step()在for loop里面

File: bluefog/torch/optimizers.py [hunk 6]
```diff
@@ -711,16 +757,21 @@ def DistributedPullGetOptimizer(optimizer, model):
...
     """An distributed optimizer that wraps another torch.optim.Optimizer with
     pull model average through bf.win_put ops.
 
     Arguments:
         optimizer: Optimizer to use for computing gradients and applying updates.
         model: The model or a list of models you want to train with.
+        num_steps_per_communication: Number of expected step() function calls before each
+                                     communication. This allows local model parameter updates
+                                     per num_steps_per_communication before reducing them over
+                                     distributed computation resources.
```
Comment (ybc1991) [inline bluefog/torch/optimizers.py:775]: Add examples of how to correctly use num_steps_per_communication in two scenarios.

Comment (hanbinhu) [reply bluefog/torch/optimizers.py:817]: Done.

File: bluefog/torch/optimizers.py (hunk 5 above)
Comment (hanbinhu) [reply bluefog/torch/optimizers.py:357]: Add the two examples below.

File: bluefog/torch/optimizers.py (hunk 4 above)
Comment (hanbinhu) [reply bluefog/torch/optimizers.py:328]: As we agreed offline, we will still trigger this at the last iteration.

File: bluefog/torch/optimizers.py (hunk 3 above)
Comment (hanbinhu) [reply bluefog/torch/optimizers.py]: Done.

File: bluefog/torch/optimizers.py (hunk 2 above)
Comment (hanbinhu) [reply bluefog/torch/optimizers.py:452]: Updated this one.

File: bluefog/torch/optimizers.py (hunk 1 above)
Comment (hanbinhu) [reply bluefog/torch/optimizers.py:325]: Yes. If model forward functions are called more than once each minibatch/batch, it may happen.

File: bluefog/torch/optimizers.py [hunk 7]
```diff
@@ -684,22 +721,73 @@ def DistributedPushSumOptimizer(optimizer, model):
     Arguments:
         optimizer: Optimizer to use for computing gradients and applying updates.
         model: The model or a list of models you want to train with.
+        num_steps_per_communication: Number of expected model forward function calls before each
+                                     communication. This allows local model parameter updates
+                                     per num_steps_per_communication before reducing them over
+                                     distributed computation resources.
+
+    Example for  two scenarios to use num_steps_per_communication:
```
Comment (ybc1991) [inline bluefog/torch/optimizers.py:729]: Extra space

File: bluefog/torch/optimizers.py [hunk 8]
```diff
@@ -684,22 +721,73 @@ def DistributedPushSumOptimizer(optimizer, model):
...
+        Scenario 1) w_{i+1} = neighbor_allredce(w_i) + \sum \grad_j(w_i)
```
Comment (ybc1991) [inline bluefog/torch/optimizers.py:730]: Would you double check the format of auto document looks fine with this format?

>>> block should have extra space

File: bluefog/torch/optimizers.py [hunk 9]
```diff
@@ -756,6 +871,31 @@ def DistributedNeighborAllreduceOptimizer(optimizer, model):
...
+                    No local model parameter update, accumulate the gradients for reducing
+        >>> opt = bf.DistributedOptimizer(optimizer, model, num_steps_per_communication=J)
```
Comment (ybc1991) [inline bluefog/torch/optimizers.py:882]: Copy paste error.

File: bluefog/torch/optimizers.py [hunk 10]
```diff
@@ -756,6 +871,31 @@ def DistributedNeighborAllreduceOptimizer(optimizer, model):
...
         optimizer: Optimizer to use for computing gradients and applying updates.
         model: The model or a list of models you want to train with.
+        num_steps_per_communication: Number of expected model forward function calls before each
+                                     communication. This allows local model parameter updates
+                                     per num_steps_per_communication before reducing them over
+                                     distributed computation resources.
+
+    Example for  two scenarios to use num_steps_per_communication:
+        Scenario 1) w_{i+1} = neighbor_allredce(w_i) + \sum \grad_j(w_i)
+                    No local model parameter update, accumulate the gradients for reducing
```
Comment (ybc1991) [inline bluefog/torch/optimizers.py:881]: Maybe just use the wording should be enough.

Scenario 1) Local accumulation of gradient without update model. (Used in large batch size or large model cases)
Scenario 2) Local updating the model. (Used in case that decreasing the communication).

File: bluefog/torch/optimizers.py [hunk 11]
```diff
@@ -756,6 +871,31 @@ def DistributedNeighborAllreduceOptimizer(optimizer, model):
...
+        >>> opt.zero_grad()
+        >>> for j in range(J):
+        >>>     output = model(data_batch_i)
+        >>>     loss = ...
+        >>>     loss.backward()
+        >>> opt.step()  # Neighbor allreducing happens here
+        Scenario 2) w_{i, j+1} = w_{i, j} - \grad_j(w_{i,j}), j = 0, 1, ....J-1
+                    w_{i+1, 0} = neighbor_allredce(w_{i, 0}）
+                    Local model parameter update
+        >>> opt = bf.DistributedOptimizer(optimizer, model, num_steps_per_communication=J)
```
Comment (ybc1991) [inline bluefog/torch/optimizers.py:892]: And many other places with copy past error.

File: bluefog/torch/optimizers.py [hunk 12]
```diff
@@ -764,16 +904,42 @@ def DistributedNeighborAllreduceOptimizer(optimizer, model):
...
+    Example for  two scenarios to use num_steps_per_communication:
+        Scenario 1) w_{i+1} = neighbor_allredce(w_i) + \sum \grad_j(w_i)
+                    No local model parameter update, accumulate the gradients for reducing
+        >>> opt = bf.DistributedOptimizer(optimizer, model, num_steps_per_communication=J)
+        >>> opt.zero_grad()
+        >>> for j in range(J):
+        >>>     output = model(data_batch_i)
+        >>>     loss = ...
+        >>>     loss.backward()
+        >>> opt.step()  # Neighbor allreducing happens here
```
Comment (ybc1991) [inline bluefog/torch/optimizers.py:932]: Same

File: bluefog/torch/optimizers.py [hunk 13]
```diff
@@ -137,7 +137,7 @@ def _timeline_forward_end_hook(module, *unused):
 
 
 class _DistributedOptimizer(torch.optim.Optimizer):
-    def __init__(self, params, model):
+    def __init__(self, params, model, num_steps_per_communication=1):
```
Comment (ybc1991) [inline bluefog/torch/optimizers.py:140]: Would you please change this back to the exact same style as Horovod?
Because we should provide two allreduce optimizers.  Following is my argument

1) like the horovod style (i.e. this one), which applied the allreduce on the gradient. In this case, it cannot have multiple local updates then communication. So  this num_steps_per_communication should change back to horovod style.

2) We can also do the same allreduce on parameter like the one we did in neighbor allreduce). In that case, num_steps_per_communication should be the same as you did in neighbor allreduce. (I am thinking to add a switch parameter to control use neighbor allreduce or allreduce in forward hook)

**Review by ybc1991 — CHANGES_REQUESTED** [review]

File: bluefog/torch/optimizers.py (hunk 13 above)
Comment (hanbinhu) [reply bluefog/torch/optimizers.py]: Done.

File: bluefog/torch/optimizers.py [hunk 14]
```diff
@@ -764,16 +1063,41 @@ def DistributedNeighborAllreduceOptimizer(optimizer, model):
...
+        >>>     loss.backward()
+        >>> opt.step()  # Allreducing happens here
+        Scenario 2) Local updating the model. (Used in case that decreasing the communication).
+        >>> opt = bf.DistributedAllreduceOptimizer(optimizer, model, num_steps_per_communication=J)
+        >>> for j in range(J):
+        >>>     output = model(data_batch_i)
+        >>>     loss = ...
+        >>>     opt.zero_grad()
+        >>>     loss.backward()
+        >>>     opt.step()  # Allreducing happens at the last iteration
```
Comment (ybc1991) [inline bluefog/torch/optimizers.py:1100]: 这一排注释是不是不对了？

File: bluefog/torch/optimizers.py [hunk 15]
```diff
@@ -741,10 +970,55 @@ def DistributedBluefogOptimizer(optimizer, model):
...
         dict(_DistributedBluefogOptimizer.__dict__),
     )
-    return cls(optimizer.param_groups, model, pull_style=False)
+    return cls(optimizer.param_groups, model, num_steps_per_communication, pull_style=False)
+
+def DistributedAllreduceOptimizer(optimizer, model,
+                                  num_steps_per_communication=1):
+    """
+    An distributed optimizer that wraps another torch.optim.Optimizer through allreduce ops.
+    The communication is triggered during forward propagation happens.
```
Comment (ybc1991) [inline bluefog/torch/optimizers.py:979]: and apply allreduce on the parameters.

File: bluefog/torch/optimizers.py [hunk 16]
```diff
@@ -741,10 +970,55 @@ def DistributedBluefogOptimizer(optimizer, model):
...
+    )
+    return cls(optimizer.param_groups, model, num_steps_per_communication)
 
 
-def DistributedNeighborAllreduceOptimizer(optimizer, model):
+def DistributedNeighborAllreduceOptimizer(optimizer, model,
+                                          num_steps_per_communication=1):
     """
     An distributed optimizer that wraps another torch.optim.Optimizer through
     neighbor_allreduce ops.
```
Comment (ybc1991) [inline bluefog/torch/optimizers.py:1024]: over parameters.

File: bluefog/torch/optimizers.py [hunk 17]
```diff
@@ -764,16 +1063,41 @@ def DistributedNeighborAllreduceOptimizer(optimizer, model):
...
-    return cls(optimizer.param_groups, model)
+    return cls(optimizer.param_groups, model, num_steps_per_communication)
 
 
-def DistributedAllreduceOptimizer(optimizer, model):
+def DistributedHorovodOptimizer(optimizer, model,
+                                backward_passes_per_step=1):
     """
     An distributed optimizer that wraps another torch.optim.Optimizer through allreduce ops.
+    The communication happens when backward propagation happens, which is the same as Horovod.
```
Comment (ybc1991) [inline bluefog/torch/optimizers.py:1073]: and allreduce applied on gradient instead of parameters

File: bluefog/torch/optimizers.py (hunk 14 above)
Comment (hanbinhu) [reply bluefog/torch/optimizers.py]: Done.

File: bluefog/torch/optimizers.py [hunk 18]
```diff
@@ -254,6 +267,117 @@ def zero_grad(self):
...
+                if self._use_timeline:
+                    # End forward computation timeline
+                    bf.timeline_end_activity(parent_name+'.'+name)
+                if p.requires_grad:
+                    if self._allreduce_delay[p] <= 0:
+                        raise AssertionError(
+                            "Unexpected behavior: forward computation were computed "
+                            "more than num_steps_per_communication times before call "
+                            "to step(). Adjust num_steps_per_communication to "
+                            "accumulate gradients locally.")
```
[the remark this answers is not in the document]
Comment (hanbinhu) [reply bluefog/torch/optimizers.py]: Done.

File: bluefog/torch/__init__.py [hunk 19]
```diff
@@ -23,6 +23,7 @@
 import torch
 from bluefog.common.util import check_extension
 from bluefog.torch.optimizers import (
+    DistributedHorovodOptimizer,
```
[the remark this answers is not in the document]
Comment (hanbinhu) [reply bluefog/torch/__init__.py]: Done.

File: examples/pytorch_benchmark.py [hunk 20]
```diff
@@ -128,6 +128,9 @@ def forward(self, x):
 elif args.dist_optimizer == 'allreduce':
     optimizer = optimizer = bf.DistributedAllreduceOptimizer(
         optimizer, model=model)
+elif args.dist_optimizer == 'horovod_allreduce':
```
[the remark this answers is not in the document]
Comment (hanbinhu) [reply examples/pytorch_benchmark.py]: Done.

File: bluefog/torch/optimizers.py (hunk 13 above)
Comment (hanbinhu) [reply bluefog/torch/optimizers.py]: Done.

**Review by ybc1991 — APPROVED** [review]
